compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,379 @@
+                }) = function.dfg[block].terminator()
+                {
+                    if left == *then_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: Some(left),
+                            block_else: None,
+                            block_exit: right,
+                        });
+                    } else if left == *else_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: None,
+                            block_else: Some(left),
+                            block_exit: right,
+                        });
+                    }
+                }
+            }
+        } else if right_successors_len == 1 && next_right == Some(left) {
+            // Right branch joins the right branch, it is a if/else statement with no then
+            // I am not sure whether this case can happen, but it is not difficult to handle it
+            let cost = block_cost(right, &function.dfg);
+            if cost < cost / 2 + jump_overhead {
+                if let Some(TerminatorInstruction::JmpIf {
+                    condition: _,
+                    then_destination,
+                    else_destination,
+                    call_stack: _,
+                }) = function.dfg[block].terminator()
+                {
+                    if right == *else_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: None,
+                            block_else: Some(right),
+                            block_exit: right,
+                        });
+                    } else if right == *then_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: Some(right),
+                            block_else: None,
+                            block_exit: right,
+                        });
+                    }
+                }
+            }
+        }
+    }
+    // A conditional exit would have exactly 2 predecessors
+    result.filter(|result| cfg.predecessors(result.block_exit).len() == 2)
+}
+
+/// Computes a cost estimate of a basic block
+/// returns u32::MAX if the block has side-effect instructions
+fn block_cost(block: BasicBlockId, dfg: &DataFlowGraph) -> u32 {
+    let mut cost: u32 = 0;
+
+    for instruction in dfg[block].instructions() {
+        let instruction_cost = match &dfg[*instruction] {
+            Instruction::Binary(binary) => {
+                match binary.operator {
+                    BinaryOp::Add { unchecked }
+                    | BinaryOp::Sub { unchecked }
+                    | BinaryOp::Mul { unchecked } => if unchecked { 1 } else { return u32::MAX },
```

Comment:
Probably want a TODO for checked operations right? e.g. the extra instructions from an `Add` overflow check are known. Is the cost we are estimating purely number of instructions?

Reply:
checked operations can fail, so they are excluded from the simplification (u32::MAX)

Reply:
I think we can differentiate between side-effectual instructions and instructions that just rely on a predicate like we do in other passes (e.g. constant folding). But this can come later, for now we can just mark a max cost. We should track wanting to re-assess this block cost estimation in an issue.